src/rdkafka_broker.c

```diff
@@ -3614,20 +3650,125 @@ rd_kafka_fetch_reply_handle (rd_kafka_broker_t *rkb,
+                                         *   ignored. If it's less than, then the message should be kept.
+                                         * 
+                                         * Note: A MessageSet comprises messages from at most one transaction, so the
+                                         * logic in step 2 is done at the message set level.
+                                         */
+
+                                        for (k = 0 ; k < AbortedTxnCnt; k++) {
+                                                int64_t pid;
+                                                int64_t first_offset;
+                                                int64_t *v;
+                                                rd_kafka_aborted_txn_start_offsets_t node, *node_ptr;
+
+                                                rd_kafka_buf_read_i64(rkbuf, &pid);
+                                                rd_kafka_buf_read_i64(rkbuf, &first_offset);
+
+                                                node.pid = pid;
+                                                node_ptr = RD_AVL_FIND(aborted_txn_offsets, &node);
+                                                if (!node_ptr) {
+                                                        node_ptr = rd_malloc(sizeof(*node_ptr));
+                                                        node_ptr->pid = pid;
+                                                        node_ptr->offsets_idx = 0;
+                                                        rd_list_init(&node_ptr->offsets, 0, NULL);
+                                                        rd_list_prealloc_elems(&node_ptr->offsets,
+                                                                sizeof(int64_t), 
+                                                                AbortedTxnCnt, 0);
+                                                        RD_AVL_INSERT(aborted_txn_offsets, node_ptr, avl_node);
+                                                }
+
+                                                v = rd_list_add(&node_ptr->offsets, NULL);
+                                                *v = first_offset;
+                                        }
+
+                                        /* 
+                                         * sort each PID txn list
+                                         *
+                                         * note: aborted_txn_list is used as a queue to traverse the avl tree.
+                                         * it's also retained to facilitate destroying it.
+                                         */
+                                        aborted_txn_list_idx = 0;
+                                        rd_list_init(&aborted_txn_list, 0, NULL);
+                                        rd_list_add(&aborted_txn_list, aborted_txn_offsets->ravl_root);
+                                        while (rd_list_cnt(&aborted_txn_list) > (int)aborted_txn_list_idx) {
```

Comment:
Why are you traversing the aborted transactions again here?
Can't this txn_list be built in the parse loop above? In `if (!node_ptr) {...`

Reply:
each list in the PID tree needs sorting. elements could alternatively be placed in the individual lists in order, but that seems less efficient / ideal AFAICT.

the txn_list was originally used only as a utility to enable the breadth first traversal of the tree, being disposed immediately after. I keep it around now on your previous suggestion that this could be reused in the AVL tree destroy, where visiting all the nodes is again needed in order to destroy the lists. if the AVL tree implementation had a free_cb like list this list wouldn't be required at all. but it doesn't.

Reply:
oh and yes, this list can be built in the above loop (great idea! much better!), but the traversal is still needed for the sorting.

Reply:
update: this is now all encapsulated up.